<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
        "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
  <meta content="text/html; charset=ISO-8859-1" http-equiv="content-type">
  <title>Edesign Help</title>
  <style type="text/css">
  body {
  background-color:white;
  color:black;
  font-size:100.01%;
  margin:0;
  min-width:41em;
  padding:5px;
  text-align:left;
  }
  td {
  font-size:0.78em;
  }
  h2 {
  padding-top:1.2em;
  }
  h3 {
  padding-top:1.2em;
  }
  div#page {
  background:#FFFFFF;
  margin:25px;
  padding:0;
  text-align:left;
  width:850px;
  }
  </style>
</head>
<body>
<div id="page">


<h2>Edesign - Help</h2>
<h2>Index of Contents</h2>
<p>
<a href="#inputHelp">1. INPUT HELP</a><br>
<a href="#outputHelp">2. OUTPUT HELP</a><br>
<a href="#calculatePenalties">3. CALCULATION OF THE PENALTY VALUE</a><br>
</p>


<h2><a name="inputHelp">1. INPUT HELP</a></h2>


<h3><a name="PRIMER_SET_LEFT_EPRIMER">Set Left primer as Eprimer</a></h3>

<p>If this is checked, Left (Forward) primer is designed as Eprimer. <br>
One thiazole orange doubly labelled thymine nucleotide will appear in a designed Left primer as "Z".
</p>


<h3><a name="PRIMER_SET_RIGHT_EPRIMER">Set Right primer as Eprimer</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_SET_LEFT_EPRIMER">Set Left primer as Eprimer</a> for the Right (Reverse) primer.</p>


<h3><a name="SEQUENCE_PRIMER">Use my own Left primer</a></h3>

<p>The sequence of a left primer to check and around which to design
right primers and optional internal probes.  Must be a substring
of <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>.
<br>
<br>
If <a href="#PRIMER_SET_LEFT_EPRIMER">Set Left primer as Eprimer</a> is checked, the sequence must contain "T", "Z" or their lower cases.
<br>
"Z" denotes the labelled nucleotide, thiazole orange doubly labelled thymine nucleotide, that must not appear more than one time in the sequence.
<br>
If "Z" is not contained, Edesign program will attempt to select appropriate position of "Z" substituting "T" in the sequence.
</p>


<h3><a name="SEQUENCE_PRIMER_REVCOMP">Use my own Right primer</a></h3>

<p>The sequence of a right primer to check and around which to
design left primers and optional internal probes.  Must be a
substring of the reverse strand of <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>.
<br>
If <a href="#PRIMER_SET_RIGHT_EPRIMER">Set Right primer as Eprimer</a> is checked, equivalent restriction of <a href="#SEQUENCE_PRIMER">Use my own Left primer</a> is applied for the Right (Reverse) primer sequence.</p>
</p>


<h3><a name="PRIMER_PICK_INTERNAL_OLIGO">Pick Internal Probe</a></h3>

<p>If this is checked, Edesign will attempt to pick an internal probe or Eprobe to detect the PCR product.
</p>


<h3><a name="PRIMER_INTERNAL_OLIGO_DIRECTION">Internal Probe Direction</a></h3>

<p>This list specifies the direction of Internal Probe
<br>
If <a href="#SEQUENCE_INTERNAL_OLIGO">Use my own Internal Probe</a> is inputted, the direction must be identical.
</p>


<h3><a name="PRIMER_INTERNAL_SET_EPROBE">Set Internal Probe as Eprobe</a></h3>

<p>If this is checked, Internal Probe is designed as Eprobe. <br>
One thiazole orange doubly labelled thymine nucleotide will appear in a designed Internal Probe as "Z".
</p>


<h3><a name="SEQUENCE_INTERNAL_OLIGO">Use my own Internal Probe</a></h3>

<p>The sequence of an internal probe to check and around which to
design left and right primers.  Must be a substring of 
<a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a> or the reverse strand of <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>.
<br>
<a href="#PRIMER_INTERNAL_OLIGO_DIRECTION">Internal Probe Direction</a> must be the same direction as the input sequence.
<br>
<br>
If <a href="#PRIMER_INTERNAL_SET_EPROBE">Set Internal Probe as Eprobe</a> is checked, the sequence must contain "T", "Z" or their lower cases.
<br>
"Z" denotes the labelled nucleotide, thiazole orange doubly labelled thymine nucleotide, that must not appear more than one time in the sequence.
<br>
If "Z" is not contained, Edesign program will attempt to select appropriate position of "Z" substituting "T" in the sequence.
</p>
</p>


<h3><a name="PRIMER_INTERNAL_GENOTYPING">Genotyping by Internal Probe [Genotyping Mode]</a></h3>

<p>If this is checked, Edesign will attempt to pick an internal probe to detect the target variant.
<br>
Target variant must be specified using "[", "]" and "/". Different variants are separated by "/".
<br>
Eg&#65294;[A/G]  [A/T/G]<br>
Not only single nucleotide variants<br>
Eg&#65294;[ATTTA/TTTTG]  [ATTTA/TA/C]<br>
Insertion and Deletion <br>
Eg&#65294;[GG/]  [/TA/C]<br>
</p>


<h3><a name="PRIMER_INTERNAL_TERMINAL_EXCLUDED_VAR">Internal Probe Terminal Region Excluded for Target Variant</a></h3>

<p>The excluded nucleotides at the 5' and 3' end of the internal probe for setting the target variant.</br>
The default is 2 nucleotides. Generally an inner nucleotide is preferable for discriminating the variation.</p>
</p>


<h3><a name="PRIMER_INTERNAL_MOD_NEAR_VAR">Internal Probe Labelling Position against Target Variant</a></h3>

<p>
This is to specify the modification position in an internal probe relative to the target variant.<br>
<br>
If "[Far] Further than" is selected, the labelled nucleotide "Z" of Eprobe is designed far from the target variant.<br>
"Z" will be designed more than the specified number of nucleotides (default: 2 nucleotides) away from the target variant.<br>
With this design, the fluorescence of Eprobe will not be decreased by a mismatch base pair at the target variant position.<br>
This design is often suitable for recognizing a peak in the derivative of the fluorescent melting curve.<br>
<br>
If "[Near] Within" is selected, the labelled nucleotide "Z" of Eprobe is designed near or on the target variant.<br>
"Z" will be designed within specified nucleotides (default: 1 nucleotide) from the target variant.<br>
With this design, the fluorescence of Eprobe will be decreased by a mismatch base pair at the target variant position.<br>
This design is suitable for recognizing the target variant by on-off detection of the fluorescence.<br>
</p>


<h3><a name="SEQUENCE_ID">Sequence Id</a></h3>

<p>An identifier that is reproduced in the output to enable you to identify the chosen primers.</p>


<h3><a name="SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE (nucleotide sequence)</a></h3>

<p>The sequence from which to choose primers.  The sequence
must be presented 5' -> 3' (i.e, in the normal way).
In general, the nucleotides may be upper or lower case, but 
lower case letters are treated 
specifically if <a href="#PRIMER_LOWERCASE_MASKING">Lowercase masking</a> is set.
The entire sequence MUST be all on a single line. 
(In other words, the sequence cannot span several lines.)
</p>


<h3><a name="SEQUENCE_INCLUDED_REGION">Included Region</a></h3>

<p>A sub-region of the given sequence in which to pick primers.  For
example, often the first dozen or so nucleotides of a sequence are
vector, and should be excluded from consideration. The value for
this parameter has the form</p>
<pre>
<tt><i>&lt;start&gt;</i></tt>,<tt><i>&lt;length&gt;</i></tt>
</pre>
<p>where <tt><i>&lt;start&gt;</i></tt> is the index of the first nucleotide to consider,
and <tt><i>&lt;length&gt;</i></tt> is the number of subsequent nucleotides in the
primer-picking region.</p>


<h3><a name="SEQUENCE_TARGET">Targets</a></h3>

<p>If one or more targets are specified then a primer pair must
flank at least one of them.  A target might be a simple sequence
repeat site (for example a CA repeat) or a single nucleotide polymorphism, 
or an exon for resequencing.  The value should be a space-separated list of</p>
<pre>
<tt><i>&lt;start&gt;</i></tt>,<tt><i>&lt;length&gt;</i></tt>
</pre>
<p>pairs where <tt><i>&lt;start&gt;</i></tt> is the index of the first nucleotide of a
target, and <tt><i>&lt;length&gt;</i></tt> is its length.</p>


See also <a href="#PRIMER_INSIDE_PENALTY">Inside Target Penalty</a>, <a href="#PRIMER_OUTSIDE_PENALTY">Outside Target Penalty.</a>


<h3><a name="SEQUENCE_EXCLUDED_REGION">Excluded Regions</a></h3>

<p>Primers and probes may not overlap any region specified in this tag.
The associated value must be a space-separated list of</p>
<pre>
<tt><i>&lt;start&gt;</i></tt>,<tt><i>&lt;length&gt;</i></tt>
</pre>
<p>pairs where <tt><i>&lt;start&gt;</i></tt> is the index of the first nucleotide of
the excluded region, and <tt><i>&lt;length&gt;</i></tt> is its length.  This tag is
useful for tasks such as excluding regions of low sequence
quality or for excluding regions containing repetitive elements
such as ALUs or LINEs.</p>


<h3><a name="SEQUENCE_PRIMER_PAIR_OK_REGION_LIST">Pair OK Region List</a></h3>

<p>

This tag allows detailed specification of possible
locations of left and right primers in primer pairs.
<br><br>
The associated value must be a semicolon-separated list of</p>

<pre>
<tt><i>&lt;left_start&gt;</i></tt>,<tt><i>&lt;left_length&gt;</i></tt>,<tt><i>&lt;right_start&gt;</i></tt>,<tt><i>&lt;right_length&gt;</i></tt>
</pre>
<p>quadruples.

The left primer must be in the region specified by 
<tt><i>&lt;left_start&gt;</i></tt>,<tt><i>&lt;left_length&gt;</i></tt> and 
the right primer must be in the region specified by
<tt><i>&lt;right_start&gt;</i></tt>,<tt><i>&lt;right_length&gt;</i></tt>.

<tt><i>&lt;left_start&gt;</i></tt> and <tt><i>&lt;left_length&gt;</i></tt> specify the location of the left primer in terms of the index of
the first nucleotide in the region and the length of the region.

<tt><i>&lt;right_start&gt;</i></tt> and <tt><i>&lt;right_length&gt;</i></tt> specify the location of the right primer in analogous fashion.

As seen in the example below, if no integers are specified for a region then the location of the
corresponding primer is not constrained.

<br><br>
Example:</p>
<pre>
Pair OK Region List=100,50,300,50 ; 900,60,, ; ,,930,100
</pre>
<p>
<br>
Specifies that there are three choices:
<br><br>
Left primer in the 50 bp region starting at position 100 AND right primer
in the 50 bp region starting at position 300
<br><br>
OR
<br><br>
Left primer in the 60 bp region starting at position 900 (and right primer anywhere)
<br><br>
OR
<br><br>
Right primer in the 100 bp region starting at position 930 (and left
primer anywhere)
</p>


<h3><a name="SEQUENCE_OVERLAP_JUNCTION_LIST">Overlap Junction List</a></h3>

<p>If this list is not empty, then
the forward OR the reverse primer must overlap one of 
these junction points by at least <a href="#PRIMER_MIN_3_PRIME_OVERLAP_OF_JUNCTION">3 Prime Junction Overlap</a> 
nucleotides at the 3' end and at least <a href="#PRIMER_MIN_5_PRIME_OVERLAP_OF_JUNCTION">5 Prime Junction Overlap</a> 
nucleotides at the 5' end.
<br><br>
In other words: The junction associated with a given position is the space immediately to the right
of that position in the template sequence on the strand given as input.
<br><br>
For example:</p>
<pre>
<code>
Overlap Junction List=20
# 1-based indexes
3 Prime Junction Overlap=4

template: atcataggccatgcctgagc^gctacgact

          ok           ...gagc^gcta-3'  (left primer)
          not ok       ...gagc^gct-3'   (left primer)
          ok           3'-ctcg^cgat...  (right pimer)
          not ok        3'-tcg^cgat...  (right primer)

5 Prime Junction Overlap=5

         ok           5'-tgagc^gccg...  (left primer)
         not ok        5'-gagc^gccg...  (left primer)
         ok           ...tgagc^gctac-5' (right primer)
         not ok       ...tgagc^gcta-5'  (right primer)

</code>
</pre>


<h3><a name="SEQUENCE_INTERNAL_EXCLUDED_REGION">Internal Probe Excluded Region</a></h3>

<p>Internal probes may not overlap any region specified by this tag.
The associated value must be a space-separated list of</p>
<pre>
<tt><i>&lt;start&gt;</i></tt>,<tt><i>&lt;length&gt;</i></tt>
</pre>
<p>pairs, where &lt;start&gt; is the index of the first nucleotide of
an excluded region, and &lt;length&gt; is its length.  Often one would
make Target regions excluded regions for internal probes.</p>


<h3><a name="SEQUENCE_FORCE_LEFT_START">Force Left Primer 5' Start</a></h3>

<p>Forces the 5' end of the left primer to be at the indicated 
position. Primers are also picked if they violate certain
constraints. The default value indicates that the start of the left primer
can be anywhere.</p>


<h3><a name="SEQUENCE_FORCE_LEFT_END">Force Left Primer 3' End</a></h3>

<p>Forces the 3' end of the left primer to be at the indicated 
position. Primers are also picked if they violate certain
constraints. The default value indicates that the end of the left primer
can be anywhere.</p>


<h3><a name="SEQUENCE_FORCE_RIGHT_START">Force Right Primer 5' Start</a></h3>

<p>Forces the 5' end of the right primer to be at the indicated 
position. Primers are also picked if they violate certain
constraints. The default value indicates that the start of the right primer
can be anywhere.</p>


<h3><a name="SEQUENCE_FORCE_RIGHT_END">Force Right Primer 3' End</a></h3>

<p>Forces the 3' end of the right primer to be at the indicated 
position. Primers are also picked if they violate certain
constraints. The default value indicates that the end of the right primer
can be anywhere.</p>


<h3><a name="PRIMER_NUM_RETURN">Number To Return</a></h3>

<p>The maximum number of primer (pairs) to return.  Primer pairs
returned are sorted by their &quot;quality&quot;, in other words by the
value of the objective function (where a lower number indicates a
better primer pair).  Caution: setting this parameter to a large
value will increase running time.</p>


<h3><a name="PRIMER_MIN_3_PRIME_OVERLAP_OF_JUNCTION">3 Prime Junction Overlap</a></h3>

<p>The 3' end of the left OR the right primer must 
overlap one of the junctions in <a href="#SEQUENCE_OVERLAP_JUNCTION_LIST">Overlap Junction List</a> by this amount.
See details in <a href="#SEQUENCE_OVERLAP_JUNCTION_LIST">Overlap Junction List</a>.
</p>


<h3><a name="PRIMER_MIN_5_PRIME_OVERLAP_OF_JUNCTION">5 Prime Junction Overlap</a></h3>

<p>The 5' end of the left OR the right primer must 
overlap one of the junctions in <a href="#SEQUENCE_OVERLAP_JUNCTION_LIST">Overlap Junction List</a> by this amount.
See details in <a href="#SEQUENCE_OVERLAP_JUNCTION_LIST">Overlap Junction List</a>.
</p>


<h3><a name="PRIMER_PRODUCT_SIZE_RANGE">Product Size Ranges</a></h3>

<p>The associated values specify the lengths of the product that the
user wants the primers to create, and is a space separated list
of elements of the form</p>
<pre>
<tt><i>&lt;x&gt;-&lt;y&gt;</i></tt>
</pre>
<p>where an <tt><i>&lt;x&gt;-&lt;y&gt;</i></tt> pair is a legal range of lengths for the
product.  For example, if one wants PCR products to be between
100 to 150 nucleotides (inclusive) then one would set this parameter to
<tt><i>100-150</i></tt>.  If one desires PCR products in either the range from
100 to 150 nucleotides or in the range from 200 to 250 nucleotides then one
would set this parameter to <tt><i>100-150 200-250</i></tt>.<br>
<br>
Edesign favors product-size ranges to the left side of the parameter string.
Edesign will return primer pairs in the first range
regardless the value of the objective function for pairs in subsequent ranges.

Only if there are an insufficient number of primers in the first
range will Edesign return primers in a subsequent range.<br>
<br>
For those with primarily a computational background,
the PCR product size is the size (in base pairs) 
of the DNA fragment that would be produced by the
PCR reaction on the given sequence template.  This
would, of course, include the primers themselves.</p>




<h3><a name="PRIMER_MIN_SIZE">Primer Size</a></h3>

<p>Minimum, Optimum, and Maximum lengths (in nucleotides) of a primer oligo. Edesign will not pick primers shorter than Min or longer than Max, and with default arguments will attempt to pick primers close in size to Opt. Min cannot be smaller than 1. Max cannot be larger than 36. (This limit is governed by maximum oligo size for which melting-temperature calculations are valid.) Min cannot be greater than Max. </p>


<h3><a name="PRIMER_INTERNAL_MIN_SIZE">Internal Probe Size</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_MIN_SIZE">Primer Size</a> for the internal probe.</p>



<h3><a name="generic_penalty_weights">Penalty Weights</a></h3>

<p>This section describes "penalty weights", which allow the user to modify the criteria that Edesign uses to select the "best" primers. There are two classes of weights: for some parameters there is a 'Lt' (less than) and a 'Gt' (greater than) weight. These are the weights that Edesign uses when the value is less or greater than (respectively) the specified optimum. The following parameters have both 'Lt' and 'Gt' weights:</p>

&nbsp;&nbsp;&nbsp;&nbsp;* Product Size
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer Size
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer Tm
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Product Tm
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer GC%
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Internal Probe Size
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Internal Probe Tm
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Internal Probe GC%

<p>The Inside Target Penalty and Outside Target Penalty are similar, except that since they relate to position they do not lend themselves to the 'Lt' and 'Gt' nomenclature.</p>

<p>For the remaining parameters the optimum is understood and the actual value can only vary in one direction from the optimum:</p>

&nbsp;&nbsp;&nbsp;&nbsp;* Primer Self Complementarity
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer 3' Self Complementarity
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer #N's
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer Mispriming Similarity
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer 3' Stability
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Internal Probe Self Complementarity
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Internal Probe 3' Self Complementarity
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Internal Probe #N's
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Internal Probe Mispriming Similarity
<br>
<br>The following weights are treated specifically:
<br>
<br>Position Penalty Weight
<br>&nbsp;&nbsp;&nbsp;&nbsp;Determines the overall weight of the position penalty in calculating the penalty for a primer.
<br>
<br>Primer Weight
<br>&nbsp;&nbsp;&nbsp;&nbsp;Determines the weight of the 2 primers penalties in calculating the primer pair penalty.
<br>
<br>Internal Probe Weight
<br>&nbsp;&nbsp;&nbsp;&nbsp;Determines the weight of the Internal Probe penalty in calculating the penalty of a primer pair plus Internal Probe.
<br>

<p>The following govern the weight given to various parameters of primer pairs.</p>

&nbsp;&nbsp;&nbsp;&nbsp;* Tm difference
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer-Primer Complementarity
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer-Primer 3' Complementarity
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer Pair Mispriming Similarity 

<p>The following govern the weight given to various parameters of primer-internal probe pairs.</p>

&nbsp;&nbsp;&nbsp;&nbsp;* Primer-Internal Probe Complementarity
<br>&nbsp;&nbsp;&nbsp;&nbsp;* Primer-Internal Probe 3' Complementarity

<p>(In case of genotyping)<br/>
The following govern the weight given to parameter for genotyping. </p>

&nbsp;&nbsp;&nbsp;&nbsp;* Tm difference by the Target Variant

<p>(In case of use of Eprobe/Eprimer)<br/>
The following govern the weight given to parameter for the labelling. </p>

&nbsp;&nbsp;&nbsp;&nbsp;* Labelling Position



<h3><a name="PRIMER_MIN_GC">Primer GC%</a></h3>

<p>Minimum, Optimum, and Maximum percentage of Gs and Cs in any primer. </p>


<h3><a name="PRIMER_INTERNAL_MIN_GC">Internal Probe GC%</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_MIN_GC">Primer GC%</a> for the internal probe.</p>


<h3><a name="PRIMER_GC_CLAMP">CG Clamp</a></h3>

<p>Require the specified number of consecutive Gs and Cs at the 3'
end of both the left and right primer.  (This parameter has no
effect on the internal probe if one is requested.)</p>


<h3><a name="PRIMER_MAX_END_GC">Max GC in Primer 3' end</a></h3>

<p>The maximum number of Gs or Cs allowed in the last five 3' 
nucleotides of a left or right primer.</p>


<h3><a name="PRIMER_MIN_TM">Primer Tm</a></h3>

<p>Minimum, Optimum, and Maximum melting temperatures (Celsius) for a primer oligo. Edesign will not pick oligos with temperatures smaller than Min or larger than Max, and with default conditions will try to pick primers with melting temperatures close to Opt.</p>
<p>For Eprimer, the effect of the dye moieties is considered as described in the paper [Kimura,Y., Hanami,T., Tanaka,Y., de Hoon,M.J.L., Soma,T., Harbers,M., Lezhava,A., Hayashizaki,Y. and Usui,K. <a href="http://pubs.acs.org/doi/abs/10.1021/bi300293d">Effect of Thiazole Orange Doubly Labelled Thymidine on DNA Duplex Formation</a>. Biochemistry, 51, 6056-6067 (2012)].</p>


<h3><a name="PRIMER_INTERNAL_MIN_TM">Internal Probe Tm</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_MIN_TM">Primer Tm</a> for the internal probe.</p>
<p>For Eprobe, the effect of the dye moieties is considered as described in the paper [Kimura,Y., Hanami,T., Tanaka,Y., de Hoon,M.J.L., Soma,T., Harbers,M., Lezhava,A., Hayashizaki,Y. and Usui,K. <a href="http://pubs.acs.org/doi/abs/10.1021/bi300293d">Effect of Thiazole Orange Doubly Labelled Thymidine on DNA Duplex Formation</a>. Biochemistry, 51, 6056-6067 (2012)].</p>


<h3><a name="PRIMER_INTERNAL_MIN_TM_VAR">Internal Probe Tm Min for Target Variant [Genotyping Mode]</a></h3>
<p>Minimum melting temperature (Celsius) for an internal probe against the Target Variant sequence. I case of genotyping by Internal Probe, Edesign will calculate the melting temperatures of internal probes against Target Variant sequence and NOT pick internal probes smaller than this parameter.</p>


<h3><a name="PRIMER_INTERNAL_WT_DIFF_TM_VAR">(Weight) Internal Probe Tm Difference by Target Variant [Genotyping Mode]</a></h3>
<p>Penalty weight for the difference of the melting temperatures of an internal probe caused by the Target Variant. 
This parameter is used in the genotyping mode. The larger the difference, the easier it is to discriminate a mismatch case from a perfect match case.
If this weight is larger than 0, higher penalty is given to smaller TM difference.</p>


<h3><a name="PRIMER_PAIR_MAX_DIFF_TM">Max Tm Difference</a></h3>

<p>Maximum acceptable (unsigned) difference between the melting
temperatures of the left and right primers.</p>


<h3><a name="PRIMER_PAIR_WT_DIFF_TM">(Weight) Tm Difference</a></h3>

<p>Penalty weight for the TM difference between the left primer and 
the right primer.</p>


<h3><a name="PRIMER_PRODUCT_MIN_TM">Product Tm</a></h3>

<p>The minimum allowed melting temperature of the amplicon.  Please
see the documentation on <a href="#PRIMER_PRODUCT_MAX_TM">PRIMER_PRODUCT_MAX_TM</a> for details.</p>





<h3><a name="PRIMER_TM_FORMULA">Table of thermodynamic parameters</a></h3>

<p>Specifies details of melting temperature calculation. 
<br>
<br>
Edesign uses the table of thermodynamic values and the method for melting temperature
calculation suggested in the paper [SantaLucia JR (1998) "A unified
view of polymer, dumbbell and oligonucleotide DNA nearest-neighbor
thermodynamics", Proc Natl Acad Sci 95:1460-65
<a href="http://dx.doi.org/10.1073/pnas.95.4.1460">http://dx.doi.org/10.1073/pnas.95.4.1460</a>].<br>
<br>
Use tag <a href="#PRIMER_SALT_CORRECTIONS">Salt Correction Formula</a>, to specify the salt correction
method for melting temperature calculation.<br>
<br>
Example of calculating the melting temperature of an oligo if
<a href="#PRIMER_TM_FORMULA">Table of thermodynamic parameters</a>=1 and <a href="#PRIMER_SALT_CORRECTIONS">Salt Correction Formula</a>=1
recommended values:<br>
<br>
primer=CGTGACGTGACGGACT<br>
<br>
Using default salt and DNA concentrations we have</p>
<pre>
Tm = deltaH/(deltaS + R*ln(C/4))
</pre>
<p>where R is the gas constant (1.987 cal/K mol) and C is the DNA 
concentration.</p>
<pre>
deltaH(predicted) =
  = dH(CG) + dH(GT) + dH(TG) + .. + dH(CT) +
     + dH(init.w.term.GC) + dH(init.w.term.AT) =

  = -10.6 + (-8.4) + (-8.5) + .. + (-7.8) + 0.1 + 2.3  =

  = -128.8 kcal/mol
</pre>
<p>where 'init.w.term GC' and 'init.w.term AT' are two
initiation parameters for duplex formation: 'initiation with
terminal GC' and 'initiation with terminal AT'</p>
<pre>
deltaS(predicted) =
  = dS(CG) + dS(GT) + dS(TG) + .. + dS(CT) +
    + dS(init.w.term.GC) + dS(init.w.term.AT) =

  = -27.2 + (-22.4) + (-22.7) + .. + (-21.0) + (-2.8) + 4.1 =
 
  = -345.2 cal/k*mol

deltaS(salt corrected) = 
  = deltaS(predicted) + 0.368*15(NN pairs)*ln(0.05M monovalent cations) =
  = -361.736

Tm = -128.800/(-361.736+1.987*ln((5*10^(-8))/4)) =
   = 323.704 K

Tm(C) = 323.704 - 273.15 = 50.554 C
</pre>


<h3><a name="PRIMER_SALT_MONOVALENT">Concentration of Monovalent Cations</a></h3>

<p>The millimolar (mM) concentration of monovalent salt cations (usually KCl) in the PCR.
Edesign uses this argument to calculate primer/probe melting
temperatures. 
Use tag <a href="#PRIMER_SALT_DIVALENT">Concentration of Divalent Cations</a> and 
<a href="#PRIMER_INTERNAL_SALT_DIVALENT">PRIMER_INTERNAL_SALT_DIVALENT</a> to specify the concentrations
of divalent cations (in which case you should also set tag <a href="#PRIMER_DNTP_CONC">Concentration of dNTPs</a> to
a reasonable value).
</p>





<h3><a name="PRIMER_SALT_DIVALENT">Concentration of Divalent Cations</a></h3>

<p>The millimolar concentration of divalent salt cations (usually MgCl<sub>2</sub>) in
the PCR. (New in v. 1.1.0, added by Maido Remm and Triinu Koressaar)<br>
<br>
Edesign converts the concentration of divalent cations to the concentration
of monovalent cations using the formula suggested in the paper [Ahsen von N,
Wittwer CT, Schutz E (2001) "Oligonucleotide Melting Temperatures under PCR
Conditions: Nearest-Neighbor Corrections for Mg<sup>2+</sup>, Deoxynucleotide Triphosphate,
and Dimethyl Sulfoxide Concentrations with Comparison to Alternative Empirical 
Formulas", Clinical Chemistry 47:1956-61 <a href="http://www.clinchem.org/cgi/content/full/47/11/1956">http://www.clinchem.org/cgi/content/full/47/11/1956</a>].</p>
<pre>
[Monovalent cations] = [Monovalent cations] + 120*(([divalent cations] - [dNTP])^0.5)
</pre>
<p>In addition, 
if the specified concentration of dNTPs (<a href="#PRIMER_DNTP_CONC">Concentration of dNTPs</a>)
is larger than the concentration of divalent cations (<a href="#PRIMER_SALT_DIVALENT">Concentration of Divalent Cations</a>)
then the effect of the divalent cations is not considered. The
concentration of dNTPs is considered in the formula above 
because some magnesium is bound by the dNTPs.
The adjusted concentration of monovalent cations is used in the
calculation of primer/probe
melting temperature, PCR product melting temperature,
and the stability of secondary structures (when <a href="#PRIMER_THERMODYNAMIC_ALIGNMENT">PRIMER_THERMODYNAMIC_ALIGNMENT</a> is 1).
If <a href="#PRIMER_SALT_DIVALENT">Concentration of Divalent Cations</a> > 0.0, be sure to
set tag <a href="#PRIMER_DNTP_CONC">Concentration of dNTPs</a> to specify the concentration of dNTPs.
</p>




<h3><a name="PRIMER_DNTP_CONC">Concentration of dNTPs</a></h3>

<p>The millimolar concentration of the sum of all deoxyribonucleotide triphosphates (dNTP). 
A reaction mix containing 0.2 mM ATP, 0.2 mM CTP, 0.2 mM GTP and 0.2 mM TTP would have a 
Concentration of dNTPs=0.8.  This argument is considered for primer/probe
primer melting temperatures, for PCR product melting temperature, or for secondary 
structure calculations only if 
<a href="#PRIMER_SALT_DIVALENT">Concentration of Divalent Cations</a> is > 0.0.
See <a href="#PRIMER_SALT_DIVALENT">Concentration of Divalent Cations</a>.</p>




<h3><a name="PRIMER_SALT_CORRECTIONS">Salt Correction Formula</a></h3>

<p>Specifies the salt correction formula for the melting temperature
calculation.  
<br>
<br>
In Edesign, only "SantaLucia 1998" is available, the salt correction formula in the paper [SantaLucia JR (1998) 
"A unified view of polymer, dumbbell and oligonucleotide DNA nearest-neighbor thermodynamics",
Proc Natl Acad Sci 95:1460-65 <a href="http://dx.doi.org/10.1073/pnas.95.4.1460">http://dx.doi.org/10.1073/pnas.95.4.1460</a>].
<br>
<br>
Edesign considers the values of the tags <a href="#PRIMER_SALT_DIVALENT">Concentration of Divalent Cations</a>, 
<a href="#PRIMER_INTERNAL_SALT_DIVALENT">PRIMER_INTERNAL_SALT_DIVALENT</a>,
<a href="#PRIMER_DNTP_CONC">Concentration of dNTPs</a>, and
<a href="#PRIMER_INTERNAL_DNTP_CONC">PRIMER_INTERNAL_DNTP_CONC</a>.
</p>


<h3><a name="PRIMER_DNA_CONC">Annealing Primer Concentration</a></h3>

<p>
<p>
A value using the nanomolar (nM) concentration of each annealing primer over the course the PCR.
Edesign uses this argument to estimate primer melting temperatures. 
This parameter corresponds to 'c' in the equation (ii) 
of the paper [SantaLucia (1998) A unified view of polymer, dumbbell, 
and oligonucleotide DNA nearest-neighbor thermodynamics. 
Proc Natl Acad Sci 95:1460-1465
<a href="http://www.pnas.org/content/95/4/1460.full.pdf+html">http://www.pnas.org/content/95/4/1460.full.pdf+html</a>],
where a suitable value (for a lower initial concentration of template) 
is "empirically determined".<br>
<br>
The default (50nM) works well with the standard
protocol used at the Whitehead/MIT Center for Genome 
Research--0.5 microliters of 20 micromolar concentration for each
primer in a 20 microliter reaction with 10 nanograms
template, 0.025 units/microliter Taq polymerase in 0.1 mM each
dNTP, 1.5mM MgCl<sub>2</sub>, 50mM KCl, 10mM Tris-HCL (pH 9.3) using 35
cycles with an annealing temperature of 56 degrees Celsius.<br>
<br>
The value of this parameter is less than the actual
concentration of primers in the initial reaction mix because it is the
concentration of annealing primers, which in turn depends on the
amount of template (including PCR product) in a given cycle.
This concentration increases a great deal during a PCR;
fortunately PCR seems quite robust for a variety of primer melting
temperatures.
<br>
See ADVICE FOR PICKING PRIMERS.</p>


<h3><a name="PRIMER_INTERNAL_DNA_CONC">Internal Probe Concentration</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_DNA_CONC">Annealing Primer Concentration</a> for the internal probe.</p>



<h3><a name="PRIMER_THERMODYNAMIC_ALIGNMENT"> Use Thermodynamic Alignment (Boolean)</a></h3>

<p>If this is checked (the associated value = 1), then Edesign will use thermodynamic models to calculate the propensity of oligos to form hairpins and dimers.</p>


<h3><a name="PRIMER_MAX_SELF_ANY">OLD: Primer Max Self Complementarity</a></h3>

<p>This parameter describes the tendency of a primer to bind to 
itself (interfering with target sequence binding). It will score 
ANY binding occurring within the entire primer sequence.<br>
It is the maximum allowable local alignment score when testing 
a single primer for (local) self-complementarity and the 
maximum allowable local alignment score when testing for 
complementarity between left and right primers.  Local 
self-complementarity is taken to predict the tendency of primers 
to anneal to each other without necessarily causing self-priming 
in the PCR.  The scoring system gives 1.00 for complementary 
bases, -0.25 for a match of any base (or N) with an N, -1.00 for 
a mismatch, and -2.00 for a gap. Only single-base-pair gaps are 
allowed. For example, the alignment</p>
<pre>
   5' ATCGNA 3'
      || | |
   3' TA-CGT 5'
</pre>
is allowed (and yields a score of 1.75), but the alignment
<pre>
   5' ATCCGNA 3'
      ||  | |
   3' TA--CGT 5'
</pre>
<p>is not considered.  Scores are non-negative, and a score of 0.00
indicates that there is no reasonable local alignment between two
oligos.</p>



<h3><a name="PRIMER_MAX_SELF_ANY_TH">TH: Primer Max Self Complementarity (decimal, 9999.99)</a></h3>

<p> The same as <a href="#PRIMER_MAX_SELF_ANY">OLD: Primer Max Self Complementarity</a> but all calculations are based on
	  thermodynamical approach. The melting temperature of the most stable
	  structure is calculated. To calculate secondary structures nearest-neighbor
	  parameters for perfect matches, single internal mismatches, terminal
	  mismatches, dangling ends have been used. Also parameters for
increments for length dependence of
bulge and internal loops have been used. This parameter is calculated only
if PRIMER_THERMODYNAMIC_ALIGNMENT=1. The by default value is 10 degrees
lower than the default value of PRIMER_MIN_TM. </p>



<h3><a name="PRIMER_INTERNAL_MAX_SELF_ANY">OLD: Internal Probe Max Self Complementarity</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_MAX_SELF_ANY">OLD: Primer Max Self Complementarity</a> for the internal probe.</p>


<h3><a name="PRIMER_INTERNAL_MAX_SELF_ANY_TH">TH: Internal Probe Max Self Complementarity (decimal, 9999.99)</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_MAX_SELF_ANY_TH">TH: Primer Max Self Complementarity</a> for the internal probe.</p>


<h3><a name="PRIMER_PAIR_MAX_COMPL_ANY">OLD: Pair Max Any Complementarity</a></h3>

<p>This parameter describes the tendency of the left primer 
to bind to the right primer or the right primer to bind to the left primer. 
It is similar to <a href="#PRIMER_MAX_SELF_ANY">OLD: Primer Max Self Complementarity</a>.</p>
<p>Edesign also calculates the complementarity of the left primer to 
the internal probe and the right primer to the internal probe.</p>


<h3><a name="PRIMER_PAIR_MAX_COMPL_ANY_TH">TH: Pair Max Any Complementarity (decimal, 9999.99)</a></h3>

<p>This parameter describes the tendency of the left primer 
    to bind to the right primer or the right primer bind to the left primer. 
	It is similar to <a href="#PRIMER_MAX_SELF_ANY_TH">TH: Primer Max Self Complementarity</a>.</p>
<p>Edesign also calculates the complementarity of the left primer to 
the internal probe and the right primer to the internal probe.</p>


<h3><a name="PRIMER_WT_SELF_ANY">(Weight) OLD: Primer Self Complementarity</a></h3>

<p>Penalty weight for the individual primer self-binding value as in 
<a href="#PRIMER_MAX_SELF_ANY">OLD: Primer Max Self Complementarity</a>.</p>


<h3><a name="PRIMER_WT_SELF_ANY_TH">(Weight) TH: Primer Self Complementarity (float)</a></h3>

<p>Penalty weight for the individual primer self-binding value as in <a href="#PRIMER_MAX_SELF_ANY_TH">TH: Primer Max Self Complementarity</a>.</p>


<h3><a name="PRIMER_INTERNAL_WT_SELF_ANY">(Weight) OLD: Internal Probe Self Complementarity</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_WT_SELF_ANY">(Weight) OLD: Primer Self Complementarity</a> for the internal probe.</p>


<h3><a name="PRIMER_INTERNAL_WT_SELF_ANY_TH">(Weight) TH: Internal Probe Self Complementarity (float)</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_WT_SELF_ANY_TH">(Weight) TH: Primer Self Complementarity </a> for the internal probe.</p>


<h3><a name="PRIMER_PAIR_WT_COMPL_ANY">(Weight) OLD: Pair Any Complementarity</a></h3>

<p>Penalty weight for the binding value of the primer pair as in <a href="#PRIMER_PAIR_MAX_COMPL_ANY">OLD: Pair Max Any Complementarity</a>.</p>


<h3><a name="PRIMER_PAIR_WT_COMPL_ANY_TH">(Weight) TH: Pair Any Complementarity (float)</a></h3>

<p>Penalty weight for the binding value of the primer pair as in <a href="#PRIMER_PAIR_MAX_COMPL_ANY_TH">TH: Pair Max Any Complementarity</a>.</p>


<h3><a name="PRIMER_MAX_SELF_END">OLD: Primer Max 3' Self Complementarity</a></h3>

<p>
Edesign calculates the binding of the primer 3'-END to an identical primer 
and scores the best binding it can find. This is critical for 
primer quality because if primers use themselves as a target 
and amplify a short piece (forming a primer-dimer), these primers 
are then unable to bind and amplify the target sequence.<br>
This parameter is the maximum allowable 3'-anchored global 
alignment score when testing a single primer for 
self-complementarity, and the maximum allowable 3'-anchored 
global alignment score when testing for complementarity between 
left and right primers. The 3'-anchored global alignment score 
is taken to predict the likelihood of PCR-priming primer-dimers, 
for example</p>
<pre>
   5' ATGCCCTAGCTTCCGGATG 3'
                ||| |||||
             3' AAGTCCTACATTTAGCCTAGT 5'
</pre>
or
<pre>
   5` AGGCTATGGGCCTCGCGA 3'
                  ||||||
               3' AGCGCTCCGGGTATCGGA 5'
</pre>
<p>The scoring system is as for the Maximum Complementarity
argument. In the examples above the scores are 7.00 and 6.00
respectively. Scores are non-negative, and a score of 0.00
indicates that there is no reasonable 3'-anchored global
alignment between two oligos.  In order to estimate 3'-anchored
global alignments for candidate primers and primer pairs, Edesign
assumes that the sequence from which to choose primers is
presented 5'->3'. It is nonsensical to provide a larger value
for this parameter than for the Maximum (local) Complementarity
parameter (<a href="#PRIMER_MAX_SELF_ANY">OLD: Primer Max Self Complementarity</a>) because the score of a local 
alignment will always be at least as great as the score of a 
global alignment.</p>



<h3><a name="PRIMER_MAX_SELF_END_TH">TH: Primer Max 3' Self Complementarity (decimal 9999.99)</a></h3>

<p>Same as <a href="#PRIMER_MAX_SELF_END">OLD: Primer Max 3' Self Complementarity</a> but is based on
thermodynamical approach - the stability of structure is analyzed. The value
of tag is expressed as melting temperature. See
<a href="#PRIMER_MAX_SELF_ANY_TH">TH: Primer Max Self Complementarity</a> for details.</p>


<h3><a name="PRIMER_INTERNAL_MAX_SELF_END">OLD: Internal Probe Max 3' End Self Complementarity</a></h3>

<p>This parameter is meaningless when internal probes have 3' end modification to prevent extension for hybridization-based detection.  
We recommend that this parameter is set at least as high as
<a href="#PRIMER_INTERNAL_MAX_SELF_ANY">OLD: Internal Probe Max Self Complementarity</a>.</p>



<h3><a name="PRIMER_INTERNAL_MAX_SELF_END_TH">TH: Internal Probe Max 3' End Self Complementarity (decimal 9999.99)</a></h3>

<p>Same as <a href="#PRIMER_INTERNAL_MAX_SELF_END">OLD: Internal Probe Max 3' End Self Complementarity</a> but for
calculating the score (melting temperature of structure) thermodynamical
approach is used.</p>


<h3><a name="PRIMER_PAIR_MAX_COMPL_END">OLD: Pair Max 3' Complementarity</a></h3>

<p>Edesign tries to bind the 3'-END of the left primer 
to the right primer or the 3'-END of the right primer to the left primer, 
and scores the best binding it can find.
It is similar to <a href="#PRIMER_MAX_SELF_END">OLD: Primer Max 3' Self Complementarity</a>.</p>
<p>Edesign also calculates the complementarity of the 3'-END of the left primer to 
the internal probe and the 3'-END of the right primer to the internal probe.</p>


<h3><a name="PRIMER_PAIR_MAX_COMPL_END_TH">TH: Pair Max 3' Complementarity (decimal, 9999.99)</a></h3>

<p>Same as <a href="#PRIMER_PAIR_MAX_COMPL_END">OLD: Pair Max 3' Complementarity</a> but for calculating the
score (melting temperature of structure) thermodynamical approach is used.</p>
<p>Edesign also calculates the complementarity of the 3'-END of the left primer to 
the internal probe and the 3'-END of the right primer to the internal probe.</p>


<h3><a name="PRIMER_WT_SELF_END">(Weight) OLD: Primer 3' Self Complementarity</a></h3>

<p>Penalty weight for the individual primer self-binding value as in 
<a href="#PRIMER_MAX_SELF_END">OLD: Primer Max 3' Self Complementarity</a>.</p>


<h3><a name="PRIMER_WT_SELF_END_TH">(Weight) TH: Primer 3' Self Complementarity (float)</a></h3>

<p>Penalty weight for the individual primer self-binding
value as in <a href="#PRIMER_MAX_SELF_END_TH">TH: Max 3' Self Complementarity</a></p>


<h3><a name="PRIMER_INTERNAL_WT_SELF_END">(Weight) OLD: Internal Probe 3' Complementarity</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_WT_SELF_END">(Weight) OLD: Primer 3' Self Complementarity</a> for the internal probe.</p>


<h3><a name="PRIMER_INTERNAL_WT_SELF_END_TH">(Weight) TH: Internal Probe 3' Complementarity (float)</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_WT_SELF_END_TH">(Weight) TH: Primer 3' Self Complementarity</a> for the internal probe.</p>


<h3><a name="PRIMER_PAIR_WT_COMPL_END">(Weight) OLD: Pair 3' Complementarity</a></h3>

<p>Penalty weight for the binding value of the primer pair as in 
<a href="#PRIMER_PAIR_MAX_COMPL_END">OLD: Pair Max 3' Complementarity</a>.</p>


<h3><a name="PRIMER_PAIR_WT_COMPL_END_TH">(Weight) TH: Pair 3' Complementarity (float)</a></h3>

<p>Penalty weight for the binding value of the primer pair as in <a href="#PRIMER_PAIR_MAX_COMPL_END_TH">TH: Pair Max 3' Complementarity</a>.</p>



<h3><a name="PRIMER_MAX_HAIRPIN_TH">TH: Primer Max Hairpin (float)</a></h3>

<p>This is the most stable monomer structure of internal probe
calculated by thermodynamic approach. The hairpin loops,
bulge loops, internal loops, internal single mismatches, dangling ends,
terminal mismatches have been considered. This parameter is calculated only
if PRIMER_THERMODYNAMIC_ALIGNMENT=1. The by default value is 10 degrees
lower than the default value of PRIMER_MIN_TM.</p>
<p>It is recommended to use low value for Eprimer.</p>


<h3><a name="PRIMER_INTERNAL_MAX_HAIRPIN_TH">TH: Internal Probe Max Hairpin (float)</a></h3>

<p>The most stable monomer structure of internal probe
calculated by thermodynamic approach. See <a href="#PRIMER_MAX_HAIRPIN_TH">TH: Primer Max Hairpin</a> for details.</p>
<p>It is recommended to use low value for Eprobe.</p>


<h3><a name="PRIMER_WT_HAIRPIN_TH">(Weight) TH: Primer Hairpin (float)</a></h3>

<p>Penalty weight for the individual primer hairpin structure value as in <a href="#PRIMER_MAX_HAIRPIN_TH">TH: Primer Max Hairpin</a>.</p>


<h3><a name="PRIMER_INTERNAL_WT_HAIRPIN_TH">(Weight) TH: Internal Probe Hairpin (float)</a></h3>

<p>Penalty weight for the most stable
primer hairpin structure value as in <a href="#PRIMER_INTERNAL_MAX_HAIRPIN_TH">TH: Internal Probe Max Hairpin</a>.</p>


<h3><a name="PRIMER_MAX_END_STABILITY">Primer Max 3' Stability</a></h3>

<p>The maximum stability for the last five 3' nucleotides of a left or
right primer.  Bigger numbers mean more stable 3' ends.  The
value is the maximum delta G (kcal/mol) for duplex disruption for
the five 3' nucleotides as calculated using the nearest-neighbor
parameter values specified by the option of <a href="#PRIMER_TM_FORMULA">Table of thermodynamic parameters</a> 
<br>
<br>
For example, if the table of thermodynamic parameters suggested
by <a href="http://dx.doi.org/10.1073/pnas.95.4.1460">SantaLucia 1998, DOI:10.1073/pnas.95.4.1460</a> is used the deltaG 
values for the most stable and for the most labile 5mer duplex 
are 6.86 kcal/mol (GCGCG) and 0.86 kcal/mol (TATAT) respectively.<br>
<!--
<br>	
If the table of thermodynamic parameters suggested by 
<a href="http://dx.doi.org/10.1073/pnas.83.11.3746">Breslauer et al. 1986, 10.1073/pnas.83.11.3746</a> is used the deltaG 
values for the most stable and for the most labile 5mer are 
13.4 kcal/mol (GCGCG) and 4.6 kcal/mol (TATAC) respectively.
-->


<h3><a name="PRIMER_WT_END_STABILITY">(Weight) Primer 3' Stability</a></h3>

<p>Penalty weight for the calculated maximum stability for the 
last five 3' nucleotides of a left or right primer.</p>


<h3><a name="PRIMER_MAX_NS_ACCEPTED">Primer Max #N's accepted</a></h3>

<p>Maximum number of unknown nucleotides (N) allowable in any primer.</p>


<h3><a name="PRIMER_INTERNAL_MAX_NS_ACCEPTED">Internal Probe Max #Ns accepted</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_MAX_NS_ACCEPTED">Primer Max #N's accepted</a> for the internal probe.</p>


<h3><a name="PRIMER_WT_NUM_NS">(Weight) Primer #N's</a></h3>

<p>Penalty weight for the number of Ns in the primer.</p>


<h3><a name="PRIMER_INTERNAL_WT_NUM_NS">(Weight) Internal Probe #N's</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_WT_NUM_NS">(Weight) Primer #N's</a> for the internal probe.</p>


<h3><a name="PRIMER_MAX_POLY_X">Primer Max Poly-X</a></h3>

<p>The maximum allowable length of a mononucleotide repeat,
for example AAAAAA.</p>


<h3><a name="PRIMER_INTERNAL_MAX_POLY_X">Internal Probe Max Poly-X</a></h3>

<p>Equivalent parameter of <a href="#PRIMER_MAX_POLY_X">Primer Max Poly-X</a> for the internal probe.</p>


<h3><a name="PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE">3' End Distance Between Left Primers</a></h3>

<p>When returning multiple primer pairs,
the minimum number of base pairs between the 3' ends of
any two left primers.<br>
<br>
Primers with 3' ends at positions e.g.  30 and 31 in the template
sequence have a three-prime distance of 1.<br>
<br>
In addition to positive values, the values -1 and 0 
are acceptable and have special interpretations:
<br>
<br>
-1 indicates that a given left primer can appear in
multiple primer pairs returned by Edesign. 
This is the default behavior.

<br>
<br>

0 indicates that a left primer is acceptable if it
was not already used.  
In other words, two left primers are allowed to
have the same 3' position provided their 5' positions differ.

<br>
<br>

For <i>n</i> &gt; 0:
A left primer is acceptable if:<br>
<br>
NOT(3' end of left primer closer than <i>n</i> to the 3' end of a previously
used left primer)<br></p>


<h3><a name="PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE">3' End Distance Between Right Primers</a></h3>

<p>Analogous to <a href="#PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE">3' End Distance Between Left Primers</a>, 
but for right primers.<br></p>


<h3><a name="PRIMER_PICK_ANYWAY">Pick anyway</a></h3>

<p>If true, use primer provided in <a href="#SEQUENCE_PRIMER">Use my own Left primer</a>, <a href="#SEQUENCE_PRIMER_REVCOMP">Use my own Right primer</a>, 
or <a href="#SEQUENCE_INTERNAL_OLIGO">Use my own Internal Probe</a> even if it violates specific 
constraints.</p>


<h3><a name="PRIMER_LOWERCASE_MASKING">Lowercase masking</a></h3>

<p>This option allows for intelligent design of primers in sequence in
which masked regions (for example repeat-masked regions) are
lower-cased. <br>
<br>
A value of 1 directs Edesign to reject primers overlapping
a lowercase nucleotide exactly at the 3' end.<br>
<br>
This property relies on the assumption that masked features
(e.g. repeats) can partly overlap primer, but they cannot overlap
the 3'-end of the primer.  In other words, lowercase nucleotides at
other positions in the primer are accepted, assuming that the
masked features do not influence the primer performance if they
do not overlap the 3'-end of primer.</p>


<h3><a name="PRIMER_EXPLAIN_FLAG">Print Statistics</a></h3>

<p>If this flag is 1 (non-0), produce <a href="#PRIMER_LEFT_EXPLAIN">PRIMER_LEFT_EXPLAIN</a>,
<a href="#PRIMER_RIGHT_EXPLAIN">PRIMER_RIGHT_EXPLAIN</a>, <a href="#PRIMER_INTERNAL_EXPLAIN">PRIMER_INTERNAL_EXPLAIN</a> and/or
<a href="#PRIMER_PAIR_EXPLAIN">PRIMER_PAIR_EXPLAIN</a> output
tags as appropriate.  
These output tags are intended to provide information on the number of
oligos and primer pairs that Edesign examined and counts of
the number discarded for various reasons.  If -format_output is
set similar information is produced in the user-oriented output.</p>


<h3><a name="PRIMER_LIBERAL_BASE">Liberal Base</a></h3>

<p>This parameter provides a quick-and-dirty way to get Edesign to
accept IUB / IUPAC codes for ambiguous nucleotides (i.e. by changing
all unrecognized nucleotides to N).  If you wish to include an
ambiguous nucleotide in an oligo, you must set <a href="#PRIMER_MAX_NS_ACCEPTED">Max #N's accepted</a> to a
1 (non-0) value.<br>
<br>
Perhaps '-' and '* ' should be squeezed out rather than changed
to 'N', but currently they simply get converted to N's.  The authors
invite user comments.</p>


<h3><a name="PRIMER_FIRST_BASE_INDEX">First Base Index</a></h3>

<p>This parameter is the index of the first nucleotide in the input
sequence.  For input and output using 1-based indexing (such as
that used in GenBank and to which many users are accustomed) set
this parameter to 1.  For input and output using 0-based indexing,
set this parameter to 0.  (This parameter also affects the
indexes in the contents of the files produced when the primer
file flag is set.)</p>


<h3><a name="PRIMER_MAX_TEMPLATE_MISPRIMING">OLD: Primer Max Template Mispriming</a></h3>

<p>The maximum allowed similarity of a primer to ectopic sites in the
template.  A negative value means do not check.  The scoring
system is the same as used for <a href="#PRIMER_MAX_LIBRARY_MISPRIMING">Primer Max Library Mispriming</a>, except
that an ambiguity code in the template is never treated as a
consensus (see <a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">Treat ambiguity codes in libraries as consensus</a></p>


<h3><a name="PRIMER_MAX_TEMPLATE_MISPRIMING_TH">TH: Primer Max Template Mispriming< (decimal, 9999.99)</a></h3>

<p>Similar to <a href="#PRIMER_MAX_TEMPLATE_MISPRIMING">OLD: Primer Max Template Mispriming</a> but 
assesses alternative binding sites in the template using thermodynamic models.
This parameter specifies the maximum allowed
melting temperature of a primer
at an "ectopic" site within the template sequence.
</p>


<h3><a name="PRIMER_INTERNAL_MAX_TEMPLATE_MISHYB">OLD: Internal Probe Max Template Mishyb</a></h3>

<p>The maximum allowed similarity of an internal probe to ectopic sites in the
template.  A negative value means do not check.  The scoring
system is the same as used for <a href="#PRIMER_MAX_LIBRARY_MISPRIMING">Primer Max Library Mispriming</a>, except
that an ambiguity code in the template is never treated as a
consensus (see <a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">Treat ambiguity codes in libraries as consensus</a></p>


<h3><a name="PRIMER_INTERNAL_MAX_TEMPLATE_MISHYB_TH">TH: Internal Probe Max Template Mishyb (decimal, 9999.99)</a></h3>

<p>Similar to <a href="#PRIMER_INTERNAL_MAX_TEMPLATE_MISHYB">OLD: Internal Probe Max Template Mishyb</a> but 
assesses alternative binding sites in the template using thermodynamic models.
This parameter specifies the maximum allowed
melting temperature of an internal probe
at an "ectopic" site within the template sequence.
</p>


<h3><a name="PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING">OLD: Pair Max Template Mispriming (decimal, 9999.99)</a></h3>

<p>The maximum allowed summed similarity of both primers to
ectopic sites in the template. A negative value means do not
check.  The scoring system is the same as used for
<a href="#PRIMER_PAIR_MAX_LIBRARY_MISPRIMING">Pair Max Library Mispriming</a>, except that an ambiguity code 
in the template is never treated as a consensus (see
<a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">Treat ambiguity codes in libraries as consensus</a>).  Edesign does not
check the similarity of internal probes to locations outside of the amplicon.</p>


<h3><a name="PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING_TH">TH: Pair Max Template Mispriming (decimal, 9999.99)</a></h3>

<p>The maximum allowed summed melting temperatures
of both primers at ectopic sites within the template (with
the two primers in an orientation that would allow PCR
amplification.) 
The melting temperatures are calculated
as for <a href="#PRIMER_MAX_TEMPLATE_MISPRIMING_TH">TH: Primer Max Template Mispriming</a>.


<h3><a name="PRIMER_WT_TEMPLATE_MISPRIMING">(Weight) OLD: Primer Template Mispriming</a></h3>

<p>Penalty weight for a single primer binding to the template sequence.</p>


<h3><a name="PRIMER_WT_TEMPLATE_MISPRIMING_TH">(Weight) TH: Primer Template Mispriming (float)</a></h3>

<p>Penalty weight for a single primer binding to the template sequence (thermodynamic approach).</p>


<h3><a name="PRIMER_INTERNAL_WT_TEMPLATE_MISHYB">(Weight) OLD: Internal Probe Template Mishyb (float)</a></h3>

<p>Penalty weight for an internal probe binding to the template sequence.</p>


<h3><a name="PRIMER_INTERNAL_WT_TEMPLATE_MISHYB_TH">(Weight) TH: Internal Probe Template Mishyb (float)</a></h3>

<p>Penalty weight for an internal probe binding to the template sequence (thermodynamic approach).</p>


<h3><a name="PRIMER_PAIR_WT_TEMPLATE_MISPRIMING">(Weight) OLD: Pair Template Mispriming</a></h3>

<p>Penalty weight for a primer pair binding to the template sequence.</p>


<h3><a name="PRIMER_PAIR_WT_TEMPLATE_MISPRIMING_TH">(Weight) TH: Pair Template Mispriming (float)</a></h3>

<p>Penalty weight for a primer pair binding to the template sequence (thermodynamic approach).</p>


<h3><a name="PRIMER_MISPRIMING_LIBRARY">Mispriming Library</a></h3>

<p>The name of a file containing a nucleotide sequence library of
sequences to avoid amplifying (for example repetitive sequences, or
possibly the sequences of genes in a gene family that should
not be amplified.)  The file must be in FASTA format.
If this parameter is specified then Edesign locally
aligns each candidate primer against each library sequence and
rejects those primers for which the local alignment score times a
specified weight (see below) exceeds <a href="#PRIMER_MAX_LIBRARY_MISPRIMING">Primer Max Library Mispriming</a>.
<br>
<br>
Each sequence entry in the FASTA-format file must begin with an
"id line" that starts with '>'.  The contents of the id line is
"slightly restricted" in that Edesign parses everything after any
optional asterisk ('*') as a floating point number to use as the
weight mentioned above.  If the id line contains no asterisk then
the weight defaults to 1.0. (The maximum value of the weight is 100.0.)<br>
The alignment scoring system used is
the same as for calculating complementarity among oligos (e.g.
<a href="#PRIMER_MAX_SELF_ANY">Max Self Complementarity</a>), except for the handling of IUB/IUPAC ambiguity
codes (discussed below).<br>
<br>
The remainder of an entry contains the sequence as lines
following the id line up until a line starting with '>' or
the end of the file.  Whitespace and newlines are ignored.
Characters 'A', 'T', 'G', 'C', 'a', 't', 'g', 'c' and
IUB/IUPAC 'ambiguity' codes ('R, 'Y', 'K', 'M', 'S', 'W',
'N', including lower case) are retained. For technical
reasons the length of the sequence must be >= 3. Of course,
sequences of length < 10 or so are probably useless, but
will be accepted without complaint.<br>
<br>
WARNING: always set <a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">Treat ambiguity codes in libraries as consensus</a>=0
if any sequence in the library contains strings of <br>
'N's: NNNNNNNNNNNNNNNNNNNN.<br>
There are no restrictions on line length.<br>
<br>
An empty value for this parameter indicates that no repeat
library should be used and "turns off" the use of a
previously specified library.<br>
<br>
The contents of the libraries can be viewed at the following links:<br>
<ul>
<li><a href="/cgi-bin/edesign/show_humrep_and_simple.cgi">HUMAN</a> (contains microsatellites)</li>
<li><a href="/cgi-bin/edesign/show_rodrep_and_simple.cgi">RODENT_AND_SIMPLE</a> (contains microsatellites)</li>
<li><a href="/cgi-bin/edesign/show_rodent_ref.cgi">RODENT</a> (does not contain microsatellites)</li>
<li><a href="/cgi-bin/edesign/show_drosophila_w_transposons.cgi">DROSOPHILA</a></li>
</ul>
</p>

<h3><a name="PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">Treat ambiguity codes in libraries as consensus</a></h3>

<p>If set to 1, treat ambiguity codes as if they were consensus
codes when matching oligos to mispriming or mishyb
libraries. For example, if this flag is set, then a C in an
oligo will be scored as a perfect match to an S in a library
sequence, as will a G in the oligo. More importantly,
though, any nucleotide in an oligo will be scored as a perfect
match to an N in the library.  This is very bad if the
library contains strings of Ns, as no oligo will be legal
(and it will take a long time to find this out). So unless
you know for sure that your library does not have runs of Ns
(or Xs), then set this flag to 0.</p>


<h3><a name="PRIMER_MAX_LIBRARY_MISPRIMING">Max Library Mispriming</a></h3>

<p>The maximum allowed weighted similarity with any sequence in
<a href="#PRIMER_MISPRIMING_LIBRARY">Mispriming Library</a>.</p>


<h3><a name="PRIMER_INTERNAL_MAX_LIBRARY_MISHYB">Internal Probe Max Library Mishyb</a></h3>

<p>Similar to <a href="#PRIMER_MAX_LIBRARY_MISPRIMING">Max Library Mispriming</a> except that this parameter applies
to the similarity of candidate internal probes to the library
specified in <a href="#PRIMER_INTERNAL_MISHYB_LIBRARY">Internal Probe Mishyb Library<</a>.</p>


<h3><a name="PRIMER_PAIR_MAX_LIBRARY_MISPRIMING">Pair Max Library Mispriming</a></h3>

<p>The maximum allowed sum of similarities of a primer pair
(one similarity for each primer) with any single sequence in
<a href="#PRIMER_MISPRIMING_LIBRARY">Mispriming Library</a>.  
Library sequence weights are not used in computing the sum
of similarities.</p>


<h3><a name="PRIMER_WT_LIBRARY_MISPRIMING">(Weight) Primer Library Mispriming</a></h3>

<p>Penalty weight for a single primer binding to any single sequence 
in <a href="#PRIMER_MISPRIMING_LIBRARY">Mispriming Library</a>.</p>


<h3><a name="PRIMER_INTERNAL_WT_LIBRARY_MISHYB">(Weight) Internal Probe Library Mishyb </a></h3>

<p>Penalty weight for an internal probe binding to any single sequence 
in <a href="#PRIMER_MISPRIMING_LIBRARY">Mispriming Library</a>.</p>


<h3><a name="PRIMER_PAIR_WT_LIBRARY_MISPRIMING">(Weight) Pair Library Mispriming</a></h3>

<p>Penalty weight for a primer pair binding to any single sequence 
in <a href="#PRIMER_MISPRIMING_LIBRARY">Mispriming Library</a>.</p>




<h3><a name="PRIMER_PAIR_WT_PR_PENALTY">(Weight) Primer Penalty Weight</a></h3>

<p>Penalty weight factor for the sum of the left and the right primer added 
to the pair penalty. Setting this value below 1.0 will increase 
running time.<br>
<br>
As <a href="#PRIMER_PAIR_WT_PR_PENALTY">(Weight) Primer Penalty Weight</a> or the per-primer 
penalties it multiplies become lower with respect to various 
pair penalties (for example <a href="#PRIMER_PAIR_WT_PRODUCT_SIZE_LT">(Weight) Product Size Lt</a>
<a href="#PRIMER_PAIR_WT_PRODUCT_SIZE_GT">(Weight) Product Size Gt</a>
<a href="#PRIMER_PAIR_WT_DIFF_TM">(Weight) Tm Difference</a>, etc.) the running time of the 
search for primer pairs is likely to grow substantially. The 
reason is that the search algorithm must calculate the penalty 
for more primer pairs.

</p>


<h3><a name="PRIMER_PAIR_WT_IO_PENALTY">(Weight) Internal Probe Penalty Weight</a></h3>

<p>Penalty weight factor for the internal probe added to the pair penalty.</p>


<h3><a name="PRIMER_INSIDE_PENALTY">Inside Target Penalty</a></h3>

<p>Non-default values are valid only for sequences with 0 or 1
target regions.  If the primer is part of a pair that spans a
target and overlaps the target, then multiply this value times
the number of nucleotide positions by which the primer overlaps
the (unique) target to get the 'position penalty'.  The effect of
this parameter is to allow Edesign to include overlap with the
target as a term in the objective function.</p>


<h3><a name="PRIMER_OUTSIDE_PENALTY">Outside Target Penalty</a></h3>

<p>Non-default values are valid only for sequences with 0 or 1
target regions.  If the primer is part of a pair that spans a
target and does not overlap the target, then multiply this value
times the number of nucleotide positions from the 3' end to the
(unique) target to get the 'position penalty'.  The effect of
this parameter is to allow Edesign to include nearness to the
target as a term in the objective function.</p>


<h3><a name="PRIMER_WT_POS_PENALTY">(Weight) Primer Position Penalty</a></h3>

<p>Penalty weight for the primer which do not overlap the target.</p>


<h3><a name="PRIMER_5_PRIME_TERMINAL_EXCLUDED_MOD">Primer 5' Terminal Region Excluded for Labelling</a></h3>

<p>The excluded nucleotides at the 5' end of the primer for modification.</br>
For Eprimer, the default is 2 nucleotides.</p>

<h3><a name="PRIMER_3_PRIME_TERMINAL_EXCLUDED_MOD">Primer 3' Terminal Region Excluded for Labelling</a></h3>

<p>The excluded nucleotides at the 3' end of the primer for modification.</br>
For Eprimer, the default is 1 nucleotide.</p>

<h3><a name="PRIMER_INTERNAL_5_PRIME_TERMINAL_EXCLUDED_MOD">Internal Probe 5' Terminal Region Excluded for Labelling</a></h3>

<p>The excluded nucleotides at the 5' end of the internal probe for modification.</br>
For Eprobe, the default is 2 nucleotides.</p>

<h3><a name="PRIMER_INTERNAL_3_PRIME_TERMINAL_EXCLUDED_MOD">Internal Probe 3' Terminal Region Excluded for Labelling</a></h3>

<p>The excluded nucleotides at the 3' end of the internal probe for modification.<br/>
For Eprobe, the default is 2 nucleotides.</p>


<h3><a name="PRIMER_INTERNAL_WT_MOD_POS">(Weight) Labelling Terminal Side</a></h3>

<p>Penalty weight for modification positions next to the excluded region.<br/>
In case those positions are possible for modification but not a best option.<br/>
<br/>
For Eprobe, the first and second nucleotide from terminal end are not preferable for labelling.<br/>
By default the excluded nucleotides for labelling are two. But labelling the third nucleotide sometimes reduces fluorescence a little.<br/>
This penalty weight is applied for the third nucleotide in this case as next to the excluded region to avoid labelling the third nucleotide.</p>



<h2><a name="outputHelp">2. OUTPUT HELP</a></h2>

<h3><a name=p3w_primer_start>start (Start Position)</a></h3>
<p>The position of the 5' nucleotide of the primer.<br>
For a Left Primer or an Internal Probe (Forward) this position is the position
of the leftmost nucleotide.<br>
For a Right primer or an Internal Probe (Reverse) it is the position of the rightmost nucleotide.</p>

<h3><a name=p3w_primer_len>len (Oligo Length)</a></h3>
<p>The length of the primer or probe.</p>

<h3><a name=p3w_primer_tm>tm (Melting Temperature)</a></h3>
<p>The melting temperature of the primer or probe.</p>

<h3><a name=p3w_primer_tm_var>tm(VAR) (Melting Temperature against VARIANT)</a></h3>
<p>The melting temperature of the primer or probe against the target variant sequence.</p>

<h3><a name=p3w_primer_gc>gc%</a></h3>
<p>The percentage of G or C nucleotides in the primer or probe.</p>

<h3><a name=p3w_primer_any>any (Self Complementarity)</a></h3>
<p>The self-complementarity score of the primer or probe (taken as
a measure of its tendency to anneal to itself or form
secondary structures).</p>

<h3><a name=p3w_primer_three>3' (Self Complementarity)</a></h3>
<p>The 3' self-complementarity of the primer or probe (taken
as a measure of its tendency to form a primer-dimer with itself).</p>

<h3><a name=p3w_primer_hairpin>hairpin (Primer or Probe Hairpin)</a></h3>
<p>The calculated value of the melting temperature of the most stable
hairpin structure of the primer or probe.

<h3><a name=p3w_primer_repeat>rep (Mispriming or Mishyb Library Similarity)</a></h3>
<p>The similarity to the specified Mispriming or Mishyb library.</p>

<h3><a name=p3w_primer_seq>seq (Primer Sequence, 5' to 3')</a></h3>
<p>The sequence of the selected primer or probe, always 5' to 3' so the
right primer is on the opposite strand from the one supplied in
the source input. (The right primer sequence is the sequence
you would want synthesized in a primer order.)
"Z" denotes the labelled nucleotide, thiazole orange doubly labelled thymine nucleotide.</p>


<h2><a name="calculatePenalties">3. CALCULATION OF THE PENALTY VALUE</a></h2>

<p>
In essence, the penalty values define what is the best primer
pair.
The calculation of penalty values
takes into consideration penalty weights, which allow
one to fine-tune the selection of primers to specific needs.<br>
<br>
This section will explain the selection process of primers by 
Edesign. In general the selection is a multi-step process:<br>
<br>
In the first step, Edesign evaluates every primer that 
can be picked in the region of interest, possibly subject to
constraints due to target regions, product size ranges, and
so forth, that might preclude the use of primers in the
eventually selected primer pairs. In this pass the hard 
limits are tested like <a href="#PRIMER_MAX_GC">PRIMER_MAX_GC</a> or <a href="#PRIMER_MIN_TM">PRIMER_MIN_TM</a>. Primers 
with a GC lower than <a href="#PRIMER_MAX_GC">PRIMER_MAX_GC</a> or a Tm higher than 
<a href="#PRIMER_MIN_TM">PRIMER_MIN_TM</a> are memorized, the primers which fail in one of 
these tests are excluded. Edesign can be forced to use primers 
failing to pass this test by setting <a href="#PRIMER_PICK_ANYWAY">Pick anyway</a> to one 
(only available for primers provided by the user).<br>
<br>
In the second step, Edesign calculates a penalty for each
primer. This penalty is the only score by which Edesign 
evaluates the primers and the internal probe. 
Edesign enhances calculation of the penalty for the internal probe.
<p>
Edesign tries to select pairs with the lowest penalty which 
still fulfill all necessary requirements like non-redundancy or
product size limits.</p>

</div>
</body>
</html>

<!--
 Copyright Notice and Disclaimer for Edesign

<p>
 Copyright (c) 2013,2014 RIKEN and K.K.DNAFORM. All Rights Reserved
 The Edesign is based on the Primer3 program (version 2.3.4) of the Whitehead Institute (http://primer3.ut.ee/).
</p>
<pre>
       This file is part of Edesign software.

       This software is free software;
       you can redistribute it and/or modify it under the terms
       of the GNU General Public License as published by the Free
       Software Foundation; either version 2 of the License, or (at
       your option) any later version.

       This software is distributed in the hope that it will be useful,
       but WITHOUT ANY WARRANTY; without even the implied warranty of
       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
       GNU General Public License for more details.

       You should have received a copy of the GNU General Public License
       along with this software (file gpl-2.0.txt in the source
       distribution); if not, write to the Free Software
       Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 OWNERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON A THEORY
 OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
</pre>


<h2>Citing Edesign</h2>
<p>
We request that use of this software be cited in
publications as
</p>
<p>
</p>

-->

